runtime/security.go

```diff
@@ -609,6 +619,372 @@ func (p *securityEngine) applySecurityRuleRowFilter(res *ResolvedSecurity, _ *ru
 	return nil
 }
 
+func (p *securityEngine) expandRules(ctx context.Context, instanceID string, claims *SecurityClaims) ([]*runtimev1.SecurityRule, error) {
+	var rules []*runtimev1.SecurityRule
+	for _, rule := range claims.AdditionalRules {
+		if rule.GetTransitiveAccess() == nil {
+			rules = append(rules, rule)
+			continue
+		}
+		// If the rule is a transitive access rule, we need to resolve it
+		resName := rule.GetTransitiveAccess().GetResource()
+		if resName == nil {
+			return nil, fmt.Errorf("transitive access rule has no resource")
+		}
+		ctr, err := p.rt.Controller(ctx, instanceID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get controller: %w", err)
+		}
+		res, err := ctr.Get(ctx, resName, false)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get resource %q of kind %q: %w", resName.Name, resName.Kind, err)
```

Comment:
Thinking about the end user experience here, imagine someone opens a report or magic link, but the resource has been deleted. They would probably expect a 404/"not found" error, not a "failed to get resource". If the resource is not found, the `ctr` would return `drivers.ErrResourceNotFound`, maybe we could add handling for that here: https://github.com/rilldata/rill/blob/7caad7ca93be914fc318727805bba718c6f1ff33/runtime/server/server.go#L334